core/trino-main/src/main/java/io/trino/server/security/oauth2/JWTTokenSerializer.java

```diff
@@ -0,0 +1,158 @@
+import static io.trino.server.security.jwt.JwtUtil.newJwtParserBuilder;
+import static io.trino.server.security.oauth2.JWTTokenSerializer.ZstdCodec.CODEC_NAME;
+import static java.lang.Math.toIntExact;
+import static java.util.Arrays.copyOfRange;
+import static java.util.Objects.requireNonNull;
+
+public class JWTTokenSerializer
+        implements TokenPairSerializer
+{
+    private static final CompressionCodec COMPRESSION_CODEC = new ZstdCodec();
+    private static final String ACCESS_TOKEN_KEY = "access_token";
+    private static final String REFRESH_TOKEN_KEY = "refresh_token";
+    private final Key signingKey;
+    private final Key verificationKey;
+    private final SignatureAlgorithm signatureAlgorithm;
+    private final OAuth2Client client;
+    private final io.jsonwebtoken.Clock clock;
+    private final String issuer;
+    private final JwtParser parser;
+
+    public JWTTokenSerializer(
+            JwtKeysProvider keysProvider,
+            SignatureAlgorithm signatureAlgorithm,
+            OAuth2Client client,
+            NodeVersion nodeVersion,
+            Clock clock,
+            long expirationTimeExtensionSeconds)
+    {
+        this.signingKey = requireNonNull(keysProvider, "keysProvider is null").getSigningKey();
+        this.verificationKey = requireNonNull(keysProvider, "keysProvider is null").getVerificationKey();
+        this.signatureAlgorithm = requireNonNull(signatureAlgorithm, "signatureAlgorithm is null");
+        this.client = requireNonNull(client, "client is null");
+        this.issuer = requireNonNull(nodeVersion, "nodeVersion is null").getVersion();
+        this.clock = () -> Date.from(requireNonNull(clock, "clock is null").instant());
+
+        this.parser = newJwtParserBuilder()
+                .setSigningKey(verificationKey)
+                .setClock(this.clock)
+                .requireIssuer(this.issuer)
+                .setCompressionCodecResolver(JWTTokenSerializer::resolveCompressionCodec)
+                .setAllowedClockSkewSeconds(expirationTimeExtensionSeconds)
+                .build();
+    }
+
+    @Override
+    public TokenPair deserialize(String token)
+    {
+        requireNonNull(token, "token is null");
+
+        try {
+            Claims claims = parser.parseClaimsJws(token).getBody();
+            return TokenPair.accessAndRefreshTokens(
+                    claims.get(ACCESS_TOKEN_KEY, String.class),
+                    claims.get(REFRESH_TOKEN_KEY, String.class));
+        }
+        catch (ExpiredJwtException ex) {
+            throw new IllegalArgumentException("Token has timed out", ex);
+        }
+    }
+
+    @Override
+    public String serialize(TokenPair tokenPair)
+    {
+        requireNonNull(tokenPair, "tokenPair is null");
+
+        Optional<Map<String, Object>> claims = client.getClaims(tokenPair.getAccessToken());
+        JwtBuilder jwt = newJwtBuilder()
+                .addClaims(claims.orElse(Map.of()))
+                .setIssuer(issuer)
```

Comment:
Ok, we should provide audience but for the subject, we would have to get this from the accessToken. I'm thinking whether it's ok, considering your previous comment. (so that would mean that we were to copy only the subject, from the original accessToken). Alternatively, we could allow configuration property as a list with claims (on top of subject) that should be copied from the accessToken - but maybe that's not required.

Reply:
Copying the subject from the AT (access token) claims is fine, but we don't want to dump all the values from the access token into the JWT we generate.

Reply:
👍